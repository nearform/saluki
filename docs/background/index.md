# Background

---

## Configuration

Configurable via;

```
background
```
