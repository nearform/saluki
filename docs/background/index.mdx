# Background

---

## Rules

```js
background.color.black
background.color.white

background.color.lighterGray
background.color.lightGray
background.color.gray
background.color.darkGray
background.color.darkerGray

background.color.lighterRed
background.color.lightRed
background.color.red
background.color.darkRed
background.color.darkerRed

background.color.lighterBlue
background.color.lightBlue
background.color.blue
background.color.darkBlue
background.color.darkerBlue

background.color.lighterYellow
background.color.lightYellow
background.color.yellow
background.color.darkYellow
background.color.darkerYellow

background.color.lighterGreen
background.color.lightGreen
background.color.green
background.color.darkGreen
background.color.darkerGreen

background.color.lighterOrange
background.color.lightOrange
background.color.orange
background.color.darkOrange
background.color.darkerOrange

background.color.lighterPurple
background.color.lightPurple
background.color.purple
background.color.darkPurple
background.color.darkerPurple

background.url()

background.size.auto
background.size.cover
background.size.contain

background.attachment.fixed
background.attachment.local
background.attachment.scroll

background.position.top
background.position.right
background.position.bottom
background.position.left
background.position.center
background.position.topRight
background.position.bottomRight
background.position.bottomLeft
background.position.topLeft

background.repeat.none
background.repeat.repeat
background.repeat.repeatX
background.repeat.repeatY
```
