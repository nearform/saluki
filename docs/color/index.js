import styled from 'styled-components/macro'
import {
  background,
  border,
  color,
  layout,
  size,
  spacing,
  text
} from '../../src'

export const Color = styled.h1`
  ${color.white}
  ${border.solid}
  ${border.blue}
  ${background.color.black}
`

export const Container = styled.div`
  ${spacing.margin.bottom.medium}
  ${layout.display.flex}
`

export const Box = styled.div`
  flex: 1;
  ${layout.display.flex}
  ${layout.justifyContent.center}
  ${layout.direction.column}
  ${({ color }) => background.color[color]}
  ${size.height.large}
`

export const Text = styled.p`
  ${text.align.center}
  ${text.weight.bold}
  ${({ light }) => light && color.white}
`
