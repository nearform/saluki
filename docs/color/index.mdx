import { Color } from './'

# Color

---

## Example

<Color>Hello</Color>

```js
const Color = styled.h1`
  ${color.white}
  ${border.solid}
  ${border.blue}
  ${background.black}
`
```

## Configuration

Configurable via;

```
color
```

## Defaults

```js
{
  black: 'black',
  blue: 'blue',
  gray: 'gray',
  green: 'green',
  orange: 'orange',
  red: 'red',
  purple: 'purple',
  yellow: 'yellow',
  white: 'white'
}
```
