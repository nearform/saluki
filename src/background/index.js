const backgroundSizes = {
  auto: 'auto',
  cover: 'cover',
  contain: 'contain'
}
const backgroundAttachments = {
  fixed: 'fixed',
  local: 'local',
  scroll: 'scroll'
}
const backgroundPositions = {
  bottom: 'bottom',
  center: 'center',
  left: 'left',
  'left-bottom': 'left bottom',
  'left-top': 'left top',
  right: 'right',
  'right-bottom': 'right bottom',
  'right-top': 'right top',
  top: 'top'
}
const backgroundRepeats = {
  'no-repeat': 'no-repeat',
  repeat: 'repeat',
  'repeat-x': 'repeat-x',
  'repeat-y': 'repeat-y'
}
const setupBackgroundColors = (background, colors) => {
  for (const [key, value] of Object.entries(colors)) {
    background[key] = {
      'background-color': value
    }
  }
}

const setupBackgroundProperties = (background, values, name, valTransform) => {
  for (const [key, value] of Object.entries(values)) {
    const val = valTransform ? valTransform(value) : value
    background[key] = {
      [`background-${name}`]: val
    }
  }
}

const setupBackground = ({sizes = {}} = {}, colors = {}) => {
  let background = {}
  setupBackgroundProperties(background, {...backgroundSizes, ...sizes}, 'size')
  setupBackgroundProperties(background, backgroundAttachments, 'attachment')
  setupBackgroundProperties(background, backgroundPositions, 'position')
  setupBackgroundProperties(background, backgroundRepeats, 'repeat')
  setupBackgroundColors(background, colors)
  
  return background
}

export default setupBackground
