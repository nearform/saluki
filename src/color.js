export const color = {
  red: 'red'
}
