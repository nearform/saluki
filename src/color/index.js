export const defaultRules = {
  black: 'black',
  white: 'white',
  lighterGray: '#d5d9dc',
  lightGray: '#9ea6ae',
  gray: '#495057',
  darkGray: '#3a4045',
  darkerGray: '#23262a',
  lighterRed: '#f9b8b8',
  lightRed: '#f25959',
  red: '#f03e3e',
  darkRed: '#a60d0d',
  darkerRed: '#470606',
  lighterBlue: '#bbdaf6',
  lightBlue: '#61a9ea',
  blue: '#1c7ed6',
  darkBlue: '#155d9e',
  darkerBlue: '#092844',
  lighterYellow: '#ffe4b3',
  lightYellow: '#ffc04d',
  yellow: '#f59f00',
  darkYellow: '#b37400',
  darkerYellow: '#4d3200',
  lighterGreen: '#c5edcc',
  lightGreen: '#77d587',
  green: '#37b24d',
  darkGreen: '#2a883b',
  darkerGreen: '#123a19',
  lighterOrange: '#fdd2b5',
  lightOrange: '#fa9551',
  orange: '#f76707',
  darkOrange: '#ae4805',
  darkerOrange: '#4a1f02',
  lighterPurple: '#e6c3ee',
  lightPurple: '#c474d8',
  purple: '#ae3ec9',
  darkPurple: '#78278b',
  darkerPurple: '#33113c'
}

const setupColor = colors => {
  let color = {}
  for (const [key, value] of Object.entries(colors)) {
    color[key] = {
      color: value
    }
  }
  return color
}

export default setupColor
