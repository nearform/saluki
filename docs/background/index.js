import styled from 'styled-components/macro'
import { background, color, size } from '../../src'

export const Color = styled.h1`
  ${color.white}
  ${background.color.blue}
`

export const Image = styled.div`
  ${size.height.large}
  ${background.url('http://placekitten.com/704/96')}
`

export const Size = styled.div`
  ${size.height.large}
  ${background.size.contain}
  ${background.url('http://placekitten.com/300')}
`

export const Attachment = styled.div`
  ${size.height.large}
  ${background.attachment.fixed}
  ${background.url('http://placekitten.com/704')}
`

export const Position = styled.div`
  ${size.height.large}
  ${background.position.right}
  ${background.repeat.none}
  ${background.url('http://placekitten.com/96')}
`

export const Repeat = styled.div`
  ${size.height.large}
  ${background.repeat.none}
  ${background.url('http://placekitten.com/96')}
`
