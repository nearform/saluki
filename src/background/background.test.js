import background, { defaultRules, combineRules } from './'
import { color } from '../color'

describe('background', () => {
  describe('when no custom config is provided', () => {
    it('should return the default config', () => {
      const combinedRules = combineRules(defaultRules, undefined, color)
      expect(background(combinedRules)).toMatchSnapshot()
    })
  })

  describe('url', () => {
    it('should create a rule using the string passed in', () => {
      const combinedRules = combineRules(defaultRules, undefined, color)
      expect(
        background(combinedRules).url('https://nearform.com/logo.png')
      ).toEqual({
        'background-image': 'url(https://nearform.com/logo.png)'
      })
    })
  })

  describe('when custom config is provided', () => {
    describe('when the custom config attempts to amend a static value', () => {
      it('should not amend the static value', () => {
        const combinedRules = combineRules(
          defaultRules,
          {
            size: { auto: 'cover' }
          },
          color
        )
        expect(background(combinedRules).size.auto['background-size']).toBe(
          'auto'
        )
      })
    })
  })
})
