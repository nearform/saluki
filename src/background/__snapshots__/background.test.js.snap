// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`background when no custom config is provided should return the default config 1`] = `
Object {
  "attachment": Object {
    "fixed": Object {
      "background-attachment": "fixed",
    },
    "local": Object {
      "background-attachment": "local",
    },
    "scroll": Object {
      "background-attachment": "scroll",
    },
  },
  "color": Object {
    "black": Object {
      "background": "black",
    },
    "blue": Object {
      "background": "#1c7ed6",
    },
    "darkBlue": Object {
      "background": "#155d9e",
    },
    "darkGray": Object {
      "background": "#3a4045",
    },
    "darkGreen": Object {
      "background": "#2a883b",
    },
    "darkOrange": Object {
      "background": "#ae4805",
    },
    "darkPurple": Object {
      "background": "#78278b",
    },
    "darkRed": Object {
      "background": "#a60d0d",
    },
    "darkYellow": Object {
      "background": "#b37400",
    },
    "darkerBlue": Object {
      "background": "#092844",
    },
    "darkerGray": Object {
      "background": "#23262a",
    },
    "darkerGreen": Object {
      "background": "#123a19",
    },
    "darkerOrange": Object {
      "background": "#4a1f02",
    },
    "darkerPurple": Object {
      "background": "#33113c",
    },
    "darkerRed": Object {
      "background": "#470606",
    },
    "darkerYellow": Object {
      "background": "#4d3200",
    },
    "gray": Object {
      "background": "#495057",
    },
    "green": Object {
      "background": "#37b24d",
    },
    "lightBlue": Object {
      "background": "#61a9ea",
    },
    "lightGray": Object {
      "background": "#9ea6ae",
    },
    "lightGreen": Object {
      "background": "#77d587",
    },
    "lightOrange": Object {
      "background": "#fa9551",
    },
    "lightPurple": Object {
      "background": "#c474d8",
    },
    "lightRed": Object {
      "background": "#f25959",
    },
    "lightYellow": Object {
      "background": "#ffc04d",
    },
    "lighterBlue": Object {
      "background": "#bbdaf6",
    },
    "lighterGray": Object {
      "background": "#d5d9dc",
    },
    "lighterGreen": Object {
      "background": "#c5edcc",
    },
    "lighterOrange": Object {
      "background": "#fdd2b5",
    },
    "lighterPurple": Object {
      "background": "#e6c3ee",
    },
    "lighterRed": Object {
      "background": "#f9b8b8",
    },
    "lighterYellow": Object {
      "background": "#ffe4b3",
    },
    "orange": Object {
      "background": "#f76707",
    },
    "purple": Object {
      "background": "#ae3ec9",
    },
    "red": Object {
      "background": "#f03e3e",
    },
    "white": Object {
      "background": "white",
    },
    "yellow": Object {
      "background": "#f59f00",
    },
  },
  "position": Object {
    "bottom": Object {
      "background-position": "bottom",
    },
    "bottomLeft": Object {
      "background-position": "left bottom",
    },
    "bottomRight": Object {
      "background-position": "right bottom",
    },
    "center": Object {
      "background-position": "center",
    },
    "left": Object {
      "background-position": "left",
    },
    "right": Object {
      "background-position": "right",
    },
    "top": Object {
      "background-position": "top",
    },
    "topLeft": Object {
      "background-position": "left top",
    },
    "topRight": Object {
      "background-position": "right top",
    },
  },
  "repeat": Object {
    "none": Object {
      "background-repeat": "no-repeat",
    },
    "repeat": Object {
      "background-repeat": "repeat",
    },
    "repeatX": Object {
      "background-repeat": "repeat-x",
    },
    "repeatY": Object {
      "background-repeat": "repeat-y",
    },
  },
  "size": Object {
    "auto": Object {
      "background-size": "auto",
    },
    "contain": Object {
      "background-size": "contain",
    },
    "cover": Object {
      "background-size": "cover",
    },
  },
  "url": [Function],
}
`;
